src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -1,30 +1,284 @@
         case _ => throw new RuntimeException("Cannot decode GetNodeData")
       }
     }
 
     val code: Int = Message.SubProtocolOffset + 0x0d
   }
 
-  case class GetNodeData(hashes: Seq[ByteString]) extends Message {
+  case class GetNodeData(commonMptHashes: Seq[ByteString]) extends Message {
     override def code: Int = GetNodeData.code
+
+    override def toString: String = {
+      s"""GetNodeData{
+         |hashes: ${commonMptHashes.map(e => Hex.toHexString(e.toArray[Byte]))}
+         |}
+       """.stripMargin
+    }
+  }
+
+  object NodeData {
+    implicit val rlpEndDec = new RLPEncoder[NodeData] with RLPDecoder[NodeData] {
+      override def encode(obj: NodeData): RLPEncodeable = {
+        import obj._
+
+        RLPList(values.map {
+          case Left(node) => RLPValue(rlpEncode(MptNode.rlpEndDec.encode(node)))
+          case Right(byteValue) => RLPValue(byteValue.toArray[Byte])
+        }: _*)
+      }
+
+      override def decode(rlp: RLPEncodeable): NodeData = rlp match {
+        case rlpList: RLPList =>
+          NodeData(rlpList.items.map { e =>
+            Try {
+              val v = rawDecode(e: Array[Byte])
+              Left(MptNode.rlpEndDec.decode(v))
+            }.getOrElse(Right(ByteString(e: Array[Byte])))
+          })
+        case _ => throw new RuntimeException("Cannot decode NodeData")
+      }
+    }
+
+    val code: Int = Message.SubProtocolOffset + 0x0e
+  }
+
+  case class NodeData(values: Seq[Either[MptNode, ByteString]]) extends Message {
+    override def code: Int = NodeData.code
+
+    override def toString: String = {
+      val v = values.map(v => v.fold(_.toString, b => s"BytString(${Hex.toHexString(b.toArray[Byte])})"))
+
+      s"""NodeData{
+         |values: $v
+         |}
+       """.stripMargin
+    }
+  }
+
+  object MptNode {
+    implicit val rlpEndDec = new RLPEncoder[MptNode] with RLPDecoder[MptNode] {
+      override def encode(obj: MptNode): RLPEncodeable = {
+        obj match {
+          case n: MptLeaf =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = true)), rlpEncode(Account.rlpEndDec.encode(value)))
+          case n: MptExtension =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), childHash.toArray[Byte])
+          case n: MptBranch =>
+            import n._
+            RLPList((childHashes :+ terminator).map(byteStringEncDec.encode): _*)
+        }
+      }
+
+      override def decode(rlp: RLPEncodeable): MptNode = rlp match {
+        case rlpList: RLPList if rlpList.items.length == 17 =>
+          MptBranch(rlpList.items.take(16).map(byteStringEncDec.decode), byteStringEncDec.decode(rlpList.items(16)))
+        case RLPList(hpEncoded, value) =>
+          hpDecode(hpEncoded: Array[Byte]) match {
+            case (decoded, true) =>
+              MptLeaf(ByteString(decoded), Account.rlpEndDec.decode(rawDecode(value)))
```

Comment:
Would it be the same to use `rlpDecode[Account](value)` (with rlpDecode being a renamed `rlp.decode[T]`) instead of `Account.rlpEndDec.decode(rawDecode(value))`?